gobblin-api/src/main/java/gobblin/records/RecordStreamWithMetadata.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package gobblin.records;
+
+import java.util.function.Function;
+
+import gobblin.source.extractor.RecordEnvelope;
+
+import io.reactivex.Flowable;
+import lombok.Data;
+
+
+/**
+ * A stream of records along with metadata (e.g. Schema).
+ * @param <D> type of record.
+ * @param <S> type of schema.
+ */
+@Data
+public class RecordStreamWithMetadata<D, S> {
+  private final Flowable<RecordEnvelope<D>> recordStream;
+  private final S schema;
```

Comment:
This is mostly replicating the current flow where the schema is only passed along once. We could move the schema to be part of the envelope, but then every record would require looking at the schema. At any rate, this design does not prevent us from putting a dummy schema here and the real schema in the envelope.

What I could see as an issue is that on detection of new schema we may want to call pass the schema through the entire flow again. In that case, even in current Gobblin, I believe we would need to re-instantiate the pipeline, as some converters have side effects on `convertSchema`.